gazebo_plugins/src/gazebo_ros_bumper.cpp

```diff
@@ -0,0 +1,226 @@
+
+/*
+ * \desc Bumper controller
+ * \author Nate Koenig
+ * \date 09 Sept. 2008
+ */
+
+#include <gazebo/plugins/ContactPlugin.hh>
+
+#include <gazebo_msgs/msg/contact_state.hpp>
+#include <gazebo_msgs/msg/contacts_state.hpp>
+#include <gazebo_plugins/gazebo_ros_bumper.hpp>
+#include <gazebo_ros/conversions/builtin_interfaces.hpp>
+#include <gazebo_ros/node.hpp>
+#include <gazebo_ros/utils.hpp>
+#include <geometry_msgs/msg/wrench.hpp>
+
+#include <memory>
+#include <string>
+
+namespace gazebo_plugins
+{
+class GazeboRosBumperPrivate
+{
+public:
+  /// Callback to be called on contact.
+  void OnUpdate();
+
+  /// A pointer to the GazeboROS node.
+  gazebo_ros::Node::SharedPtr ros_node_{nullptr};
+
+  /// Contact mesage publisher.
+  rclcpp::Publisher<gazebo_msgs::msg::ContactsState>::SharedPtr pub_{nullptr};
+
+  /// Pointer to sensor
+  gazebo::sensors::ContactSensorPtr parent_sensor_;
+
+  /// Frame name, to be used by TF.
+  std::string frame_name_;
+
+  /// Connects to pre-render events.
+  gazebo::event::ConnectionPtr update_connection_;
+};
+
+GazeboRosBumper::GazeboRosBumper()
+: impl_(std::make_unique<GazeboRosBumperPrivate>())
+{
+}
+
+GazeboRosBumper::~GazeboRosBumper()
+{
+  impl_->ros_node_.reset();
+}
+
+void GazeboRosBumper::Load(gazebo::sensors::SensorPtr _sensor, sdf::ElementPtr _sdf)
+{
+  // Initialize ROS node
+  impl_->ros_node_ = gazebo_ros::Node::Get(_sdf);
+
+  impl_->parent_sensor_ = std::dynamic_pointer_cast<gazebo::sensors::ContactSensor>(_sensor);
+  if (!impl_->parent_sensor_) {
+    RCLCPP_ERROR(impl_->ros_node_->get_logger(),
+      "Contact sensor parent is not of type ContactSensor. Aborting");
+    impl_->ros_node_.reset();
+    return;
+  }
+
+  // Contact state publisher
+  impl_->pub_ = impl_->ros_node_->create_publisher<gazebo_msgs::msg::ContactsState>(
+    "bumper_states", rclcpp::SensorDataQoS());
+
+  RCLCPP_INFO(impl_->ros_node_->get_logger(), "Publishing contact states to [%s]",
+    impl_->pub_->get_topic_name());
+
+  // Get tf frame for output
+  impl_->frame_name_ = _sdf->Get<std::string>("frame_name", "world").first;
+
+  impl_->update_connection_ = impl_->parent_sensor_->ConnectUpdated(
+    std::bind(&GazeboRosBumperPrivate::OnUpdate, impl_.get()));
+
+  impl_->parent_sensor_->SetActive(true);
+}
+
+void GazeboRosBumperPrivate::OnUpdate()
+{
+  gazebo::msgs::Contacts contacts;
+  contacts = parent_sensor_->Contacts();
+
+  gazebo_msgs::msg::ContactsState contact_state_msg;
+  contact_state_msg.header.frame_id = frame_name_;
+  contact_state_msg.header.stamp =
+    rclcpp::Time(contacts.time().sec(), static_cast<uint32_t>(contacts.time().nsec()));
+
+  ignition::math::Pose3d pose, frame_pose;
+  ignition::math::Quaterniond rot, frame_rot;
+  ignition::math::Vector3d pos, frame_pos;
+
+  frame_pos = ignition::math::Vector3d(0, 0, 0);
+  frame_rot = ignition::math::Quaterniond(1, 0, 0, 0);  // gazebo u,x,y,z == identity
+  frame_pose = ignition::math::Pose3d(frame_pos, frame_rot);
```

Comment:
It's not necessary to zero these, because the Ignition math types are automatically initialized to zero / identity.

But in any case, these variables are not needed since we always use the world frame. `rot`, `pos`, `pose` and `frame_pose` are never used.  `frame_rot` and `frame_pos` are used, but since they're always the identity, they don't make a difference and can be removed.